Fix DISTINCT for Distributed and optimize_skip_unused_shards

azat opened the pull request:
Fixes: #9046 

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix DISTINCT for Distributed when `optimize_skip_unused_shards` is set.

Detailed description / Documentation draft:
In case of optimize_skip_unused_shards the getQueryProcessingStage() can
return QueryProcessingStage::WithMergeableState because the whole
cluster has multiple nodes, but optimize_skip_unused_shards can trim it
to 1 and in this case we need final merge that won't be done with
QueryProcessingStage::WithMergeableState only with
QueryProcessingStage::Complete.

It is better to return Complete only if there is actually only one
shard, but getQueryProcessingStage() does not accept the query to
analyze it and even if it will then we will do "costly" calculations
(for optimize_skip_unused_shards) twice which is no good (and besides I
guess most of the optimize_skip_unused_shards usage will reduce number
of shards to one).

azat committed:
Pass query to the IStorage::getQueryProcessingStage()

Will be used to return correct stage with optimize_skip_unused_shards.

azat committed:
Cover DISTINCT for Distributed with one shard (test from #9046)

v2: cover case when getQueryProcessingStage() returns Complete when it
should not (pointed by @alexey-milovidov in
https://github.com/ClickHouse/ClickHouse/pull/9808#issuecomment-602291996)

alexey-milovidov commented:
But what will happen if `optimize_skip_unused_shards` is set but the number of shards after filtering is still more than one?

alexey-milovidov commented:
If we're asking remote servers to process the query to "Complete" stage, we cannot merge their results...

azat commented:
>If we're asking remote servers to process the query to "Complete" stage, we cannot merge their results...

Hm, didn't though about it, so maybe you have ideas except for filtering unused shards in the `getQueryProcessingStage`?

alexey-milovidov commented:
Maybe it's possible to return WithMergeableState even if only one shard remains after filtering but ensure that the data will be correctly processed up to Complete stage by InterpreterSelectQuery.

alexey-milovidov commented:
> except for filtering unused shards in the getQueryProcessingStage

Why not?

azat commented:
>Maybe it's possible to return WithMergeableState even if only one shard remains after filtering but ensure that the data will be correctly processed up to Complete stage by InterpreterSelectQuery.

That was my first thought, so I changed [it](https://github.com/ClickHouse/ClickHouse/blob/9a2760d9143e7dd0c79b02f2af5cfdad52b49eda/dbms/src/Interpreters/InterpreterSelectQuery.cpp#L974) (by adding `storage && storage->isRemote`) to run `executeDistinct` for remote storage always, but does not looks better, or does it?

azat committed:
StorageDistributed: cleanup skipping unused shards

azat committed:
Fix DISTINCT for Distributed and optimize_skip_unused_shards

In case of optimize_skip_unused_shards the getQueryProcessingStage() can
return QueryProcessingStage::WithMergeableState because the whole
cluster has multiple nodes, but optimize_skip_unused_shards can trim it
to 1 and in this case we need final merge that won't be done with
QueryProcessingStage::WithMergeableState only with
QueryProcessingStage::Complete.

azat commented:
>Why not?

Ok, done

>Maybe it's possible to return WithMergeableState even if only one shard remains after filtering but ensure that the data will be correctly processed up to Complete stage by InterpreterSelectQuery.

[test](https://github.com/ClickHouse/ClickHouse/pull/9808/files#diff-28bc9246a350625c1b65ca11df60d02aR14)

azat committed:
Use overload over default value for IStorage::getQueryProcessingStage()

Since clang-tidy complains about it [1]:

    2020-03-24 09:00:53 /build/obj-x86_64-linux-gnu/../dbms/src/Storages/StorageMerge.cpp:139:42: error: default arguments on virtual or override methods are prohibited [google-default-arguments,-warnings-as-errors]
    2020-03-24 09:00:53 QueryProcessingStage::Enum StorageMerge::getQueryProcessingStage(const Context & context, const ASTPtr & query_ptr) const

  [1]: https://clickhouse-builds.s3.yandex.net/9808/b8614de64af3e65c83ee35588082df373959a737/build_log_637307151_1585043194.txt

And indeed, this is a good idea!

azat commented:
Test failures doesn't looks related

alexey-milovidov committed:
Update IStorage.h

alexey-milovidov committed:
Update StorageDistributed.cpp

azat commented:
Actually this is not only fixes DISTINCT, but also makes auto-`distributed_group_by_no_merge`, which should improve the query processing

alexey-milovidov commented:
That's awesome!
